Developers Guide (#69)
AY2122S1-CS2103T-F13-3/tp: docs/DeveloperGuide.md

**Target user profile**:

* has a need to manage a significant number of contacts
* prefer desktop apps over other types
* can type fast
* prefers typing to mouse interactions
* is reasonably comfortable using CLI apps
A private 1-to-1 home tuition teacher that:

**Value proposition**: manage contacts faster than a typical mouse/GUI driven app
1. has a need to manage not more than 50 student contacts
2. has a need to manage admin details of each student, namely, lesson fees and payment
3. has a need to keep track of lesson details for each student
4. teaches lessons anytime between 8am and 10pm, on any day of the week
5. has a need to manage a busy schedule of not more than 50 lessons
6. prefers desktop apps over other types
7. can type fast
8. prefers typing to mouse interactions
9. is reasonably comfortable using CLI apps

**Value proposition**: Use **TAB** to manage customer contacts and customer relationships faster than 
a typical mouse/GUI driven app. **TAB** effortlessly keeps track of large amounts of necessary 
student and lesson information, to empower tutors to provide the best quality home tuition service.  

### User stories

> Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unlikely to have) - `*`

Author: Right, I might've forgotten

  * 1a1. TAB displays a filtered list of students with the specified tag
  
    Use case ends.
  
* 1b. User requests for a list of students that match a specified name or address value

  * 1b1. TAB displays a list of students with matching name or address

    Use case ends.

<br/>

#### Use Case: UC2 - Add a New Student

**MSS**

1. User requests to add a new student with specified fields
2. TAB adds the student with the specified fields

   Use case ends.

**Extension**

* 1a. TAB detects that the entered student data does not satisfy requirements

    * 1a1. TAB displays an error message and the recommended changes

      Use case resumes at step 2.

* 1b. TAB detects a duplicate student

    * 1b1. TAB displays an error message

      Use case ends.

<br/>

#### Use case: UC3 – Edit a Student

**MSS:**

1. User [<ins>gets a list of students (UC1)</ins>](Use-case:-UC1---Get-a-list-of-students)
2. User requests to edit certain fields of specific student in the list
3. TAB updates the corresponding fields of the student

   Use case ends.

**Extensions**

* 1a. The list is empty

  Use case ends.

* 2a. The given index is invalid

    * 2a1. TAB displays an error message

>       Use case resumes at step 1.

Author: IMO, the list is always there, so you get a list whether you like it or not. I understand that the user shouldn't request a list all over again (UC-1 step 1), but TAB will still display the list (UC-1 step 2) after a bad request. But because they are both part of the included use case (UC-1), I can only say that it starts again at step 1

>       Use case resumes at step 1.

Author: The way AB3 does it:
![image](https://user-images.githubusercontent.com/70138507/134840554-206a5d5a-3181-40c1-8133-af0e81156712.png)

* 2b. No fields to be edited are specified

    * 2b1. TAB shows an error message

      Use case resumes at step 1.

* 2c. TAB detects that the entered student data does not satisfy requirements

    * 2c1. TAB displays an error message and the recommended changes

      Use case resumes at step 1.

> * 2d. TAB detects a duplicate student

Reviewer: Since you can edit the name as well, you might end up with a student with the same name (and other fields) as an existing one in the address book.

    * 2d1. TAB shows an error message

      Use case ends.

<br/>

> ####**Use case: UC4 - Delete a Student**

Reviewer: Is there supposed to be a space between "####"? Also, for this use case, there's the use of ** ... ** but the others don't seem to have this?

      Use case resumes at step 1.

<br/>

#### Use case: UC5 – Edit Student’s Lesson Information

**MSS**

1. User [<ins>gets a list of students (UC1)</ins>](Use-case:-UC1---Get-a-list-of-students)
2. User requests to edit certain fields of a specific lesson of a specific student in the list
3. TAB updates the corresponding fields of the lesson for the specified student

   Use case ends.

**Extensions**

* 1a. The list is empty

  Use case ends.

* 2a. The specified student has no lessons

  Use case ends.

* 2b. The given index of the student is invalid

    * 2b1. TAB displays an error message

      Use case resumes at step 1.

* 2c. The given index of the lesson of the student is invalid

    * 2b1. TAB displays an error message

      Use case resumes at step 1.

* 2b. No lesson fields to be edited are specified

    * 2b1. TAB displays an error message

      Use case resumes at step 1.

* 2d. TAB detects that the entered lesson data does not satisfy requirements

    * 2d1. TAB displays an error message and the recommended changes

      Use case resumes at step 1.

* 2e. TAB detects clashing lessons

    * 2e1. TAB displays an error message

> 

Reviewer: Can I clarify if this is a feature we would be adding or do you mean TAB detects duplicate lessons?

Author: Yup this is the duplicate lessons feature, except i think clashing lessons is a more correct description of what that feature is

Reviewer: From my understanding, clashing lessons can mean that the time range overlaps each other, rather than being "equal", is that similar to your view?

Author: Yep, overlapping time range

2. TAB undoes the previous modification

   Use case ends.

**Extension**

* 1a. There are no previous modifications

  * 1a1. TAB shows an error message
  
    Use case ends.

<br/>

#### Use Case: UC10 - Redo an Undone Modification

**MSS**

1. User requests to redo a modification to TAB's data that was previously undone within the session
2. TAB reverts the previous undo command

   Use case ends.

**Extension**

* 1a. There are no previous undo requests within the session

    * 1a1. TAB shows an error message
  
      Use case ends.
    
<br/>

### Non-Functional Requirements

1. Should work on any _mainstream OS_ as long as it has Java `11` or above installed
2. The response to any user action should become visible within 2 seconds
3. Should be able to hold up to 50 students and 50 lessons without a noticeable sluggishness in performance for typical usage
4. A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse
5. The source code should be open source
6. The product is offered as a free offline service

> 7. The product should not require any internet connectivity

Reviewer: Does this mean essential features do not require internet connectivity and the "online user guide page" (line 556) is a non-essential feature?

Author: Gd point. I'll take out this NFR

**Target user profile**:

A private 1-to-1 home tuition teacher that:

1. has a need to manage not more than 50 student contacts
2. has a need to manage admin details of each student, namely, lesson fees and payment
3. has a need to keep track of lesson details for each student

> 4. teaches lessons anytime between 8am and 10pm, on any day of the week

Reviewer: ```suggestion
4. teaches lessons anytime between 8am and 10pm, GMT+8,  on any day of the week
```
or Singapore Standard Time (SST)

Author: added

5. has a need to manage a busy schedule of not more than 50 lessons
6. prefers desktop apps over other types
7. can type fast
8. prefers typing to mouse interactions
9. is reasonably comfortable using CLI apps

> 

Reviewer 2: ```suggestion
10. Has tutees belonging to Singapore’s mainstream education systems 
```

Reviewer 2: I added this line in the draft DG on teams since I defined acad stream and acad level to be mainstream tracks in Singapore.

Author: added

**Value proposition**: Use **TAB** to manage customer contacts and customer relationships faster than 
a typical mouse/GUI driven app. **TAB** effortlessly keeps track of large amounts of necessary 
student and lesson information, to empower tutors to provide the best quality home tuition service.  

### User stories

> We categorise our user stories into four main epics:

Reviewer 3: ```suggestion
We categorise our user stories into three main epics:
```

| L3  | `MEDIUM` | user                                       | edit a student's lesson                    | update lesson information                                              |
| L4  | `MEDIUM` | user                                       | add makeup lessons for a student           | differentiate between the two different lessons                        |
| L5  | `MEDIUM` | user                                       | view the lessons I have for the week       | easily manage my schedule                                              |
| L6  | `MEDIUM` | user                                       | view a calendar of my scheduled lessons    | plan ahead while managing my schedule                                  |
| U1  | `HIGH`   | new user                                   | see usage instructions                     | refer to instructions whenever I need guidance on how to use **TAB**   |
| U2  | `HIGH`   | new user                                   | purge all current data                     | remove all sample/experimental data I used while exploring **TAB**     |

> | U3  | `HIGH`   | user                                       | find a student by name                     | locate details of persons without having to go through the entire list |

Reviewer 3: There should be another row for find a student by student details?

Author: I'll change the user story for finding by name to finding by keyword, and the corresponding UC

| U4  | `HIGH`   | user                                       | undo accidental changes                    |                                                                        |

> | U5  | `MEDIUM` | user                                       | filter students by tag                     | view a list of all the students that feature a specific tag            |

Reviewer 3: ```suggestion
| U5  | `MEDIUM` | user                                       | filter students by tags                     | view a list of all the students that feature the specified tags            |
```

| U6  | `MEDIUM` | user with incomplete information           | skip less important information to add     | update later when I have the rest of the information                   |
| U7  | `MEDIUM` | user                                       | select and copy data to desktop clipboard  | paste and use the data outside of **TAB**                              |
| U8  | `LOW`    | user who uses many tags                    | view all tags                              | remember my tags easily                                                |
| U9  | `LOW`    | expert user                                | customise the commands and input fields    | choose not to follow the given template                                |

> | U10 | `LOW`    | user                                       | be reminded to update data after a lesson  | I can ensure that data stays up to date                                |

Reviewer 3: ```suggestion
| U10 | `LOW`    | user                                       | be reminded to update data after a lesson  | ensure that data stays up to date                                |
```

1. Should work on any _mainstream OS_ as long as it has Java `11` or above installed
2. The response to any user action should become visible within 2 seconds
3. Should be able to hold up to 50 students and 50 lessons without a noticeable sluggishness in performance for typical usage
4. A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse
5. The source code should be open source
6. The product is offered as a free service
7. This product is for individual use
8. Student and lesson data should be persistent
9. All features should be easily testable
10. Should be portable, i.e. user can transfer data from one device to another easily
11. Should allow users to easily navigate the app interface

### Glossary

* **Academic level**: Mainstream academic years from primary to tertiary education in Singapore (i.e. Primary 1-6, Secondary 1-5, Junior College 1-2 and Year 1-6)
* **Academic stream**: Mainstream tracks in Singapore (i.e. Express, NA, NT, IP, IB) as well as other common exam streams (e.g. IELTS, SAT, ACT)

> * **Makeup lesson**: A lesson that occurs only once, for a student who has missed a previous lesson

Reviewer 3: can you put the lessons together

Reviewer 3: can move mainstream os down also with utc+8 also

| L1  | `HIGH`   | user                                       | schedule recurring lessons for a student   | keep track of when the student's lessons are                           |
| L2  | `HIGH`   | user                                       | delete a student's lesson                  | remove lessons that I no longer have                                   |
| L3  | `MEDIUM` | user                                       | edit a student's lesson                    | update lesson information                                              |
| L4  | `MEDIUM` | user                                       | add makeup lessons for a student           | differentiate between the two different lessons                        |
| L5  | `MEDIUM` | user                                       | view the lessons I have for the week       | easily manage my schedule                                              |
| L6  | `MEDIUM` | user                                       | view a calendar of my scheduled lessons    | plan ahead while managing my schedule                                  |

> | U1  | `HIGH`   | new user                                   | see usage instructions                     | refer to instructions when I forget how to use **TAB**                 |

Reviewer: "forget how to use TAB" sounds a little strange since the classification is for "new user" not "forgetful user". What do you think of replacing this with "refer to instructions whenever I need guidance on how to use TAB"?

Author: That's a great suggestion, I'll replace it.

| U3  | `HIGH`   | user                                       | find a student by name                     | locate details of persons without having to go through the entire list |
| U4  | `HIGH`   | user                                       | undo accidental changes                    |                                                                        |
| U5  | `MEDIUM` | user                                       | filter students by tag                     | view a list of all the students that feature a specific tag            |
| U6  | `MEDIUM` | user with incomplete information           | skip less important information to add     | update later when I have the rest of the information                   |
| U7  | `MEDIUM` | user                                       | select and copy data to desktop clipboard  | paste and use the data outside of **TAB**                              |
| U8  | `LOW`    | user who uses many tags                    | view all tags                              | remember my tags easily                                                |

> | U9  | `LOW`    | expert user                                | customise the commands and input fields    | be not limited by a template                                           |

Reviewer: Minor comment; I think "so that I can _not be_ limited by a template" might sound better?
